### DISCUSSION TITLE
nvim-dap-go remote debug with `dlv --headless` using launch.json (not working out of the box)

### DISCUSSION THREAD

**Discussion by ppenguin** [NotAShelf/nvf#414] [Q&A]
I was recently having trouble with a remote debug session. The process:

1. start _debugee_ from the cli with `dlv --listen=127.0.0.1:2345 --headless --api-version=2 exec myprog  --- [lots of prog arguments]`
2. `:DapLoadLaunchJSON`
3. `\dc`  (which is `:DapContinue`) => correctly shows a list with debug options including from `launch.json`
4. Select entry containing the following setup:
    ```
    {
      "name": "Attach to remote delve",
      "type": "go",
      "request": "attach",
      "mode": "remote",
      "port": 2345,
      "host": "127.0.0.1",
      "showLog": true,
      "trace": "log",
      "logOutput": "rpc"
    },
    ```
5. Get error "debugger not found"

If however **before step (3)** I do:
```
:lua require("dap").adapters.go = { type = "server", host = "127.0.0.1", port = 2345 }
```
it works.
Obviously I expect that `dap-go` sets up the adapter according to the selected `launch.json` entry.
How to achieve this?

**Comment by horriblename** [comment]
"type" should probably be the name of the adapter and not the language itself

DAP config is split into `dap.adapters.*` and `dap.configurations.*`. `adapters` are configs of the debugger program, and `configurations` bind a filetype to adapters.

I'm not really sure what the adapter name is (it's set by dap-go plugin) but you can find them in `require('dap').adapters`

**Reply by ppenguin** [reply]
Thanks, sorry for being a bit slow, but in this case `adapters.`**`go`** is the same as the language name? So it might well be the (correct) one from `launch.json`? But still the **adapter config** as I understand is set to `type = server` by my manual action that makes things work, which would be ok for a remote `dlv --headless` session, but for other cases not, which means one couldn't set it statically anyway.

I guess what I'm getting at is that apparently the action of choosing a config from the `launch.json` must in this case affect **both** the debugger **and** the adapter config, which seems to be not working correctly? And this action must be performed by the `dap` plugin, so the logical place to look then should be the `nvim-dap` config, but where/how?

**Reply by horriblename** [reply]
I just checked, the adapter definition from dap-go is this:

```lua
{
  executable = {
    args = { "dap", "-l", "127.0.0.1:${port}" },
    command = "/nix/store/dmdz96wkxav8x5jayr6hbcxfm043wk07-delve-1.23.0/bin/dlv",
    detached = true
  },
  options = {
    initialize_timeout_sec = 20
  },
  port = "${port}",
  type = "server"
}
```

This kinda rules out the possibility of `type = server` being sufficient to solve the issue. My knowledge of DAP stuff ends here though, and I can't help you further. Maybe nvim-dap and dap-go folks can help answer this better

**Reply by horriblename** [reply]
according to dap-go [docs](https://github.com/leoluz/nvim-dap-go/blob/6aa88167ea1224bcef578e8c7160fe8afbb44848/doc/nvim-dap-go.txt#L191)

this should work:

        {
          type = "go",
          name = "Attach remote",
          mode = "remote",
          request = "attach",
        }

can you try it? convert to json and put it in launch.json

**Reply by ppenguin** [reply]
Thanks a lot for your conscientious reply. I'm pretty much AFK for a few days but I'll try asap and get back... Cheers

**Reply by ppenguin** [reply]
> this should work:
> 
> ```
>     {
>       type = "go",
>       name = "Attach remote",
>       mode = "remote",
>       request = "attach",
>     }
> ```
> 
> can you try it? convert to json and put it in launch.json

That's essentially what I had in my OP (ignoring some explicit default values); I tried it again with this in `launch.json` but it's the same result.
